### PR TITLE
[LINALG][TOSA][MHLO] Add e2e support for aten bitwise ops

### PR DESCRIPTION
**Pull request by li-plus**
Supported `AtenBitwiseAndTensorOp`, `AtenBitwiseOrTensorOp`, `AtenBitwiseXorTensorOp` and `AtenBitwiseNotOp` on multiple backend dialects. Some ops were already supported, so this PR added missing bitwise ops for these backends.

### CODE REVIEW

File: lib/Conversion/TorchToMhlo/Basic.cpp [hunk 1]
```diff
@@ -1474,5 +1493,7 @@ void mlir::torch::torch_to_mhlo::populateBasicOpPatternsAndLegality(
   INSERT_BINARY_BROADCAST_PATTERN(AtenMinimumOp, chlo::BroadcastMinOp);
   INSERT_BINARY_BROADCAST_PATTERN(Aten__And__TensorOp, chlo::BroadcastAndOp);
   INSERT_BINARY_BROADCAST_PATTERN(AtenBitwiseAndTensorOp, chlo::BroadcastAndOp);
+  INSERT_BINARY_BROADCAST_PATTERN(AtenBitwiseOrTensorOp, chlo::BroadcastOrOp);
+  INSERT_BINARY_BROADCAST_PATTERN(AtenBitwiseXorTensorOp, chlo::BroadcastXorOp);
```
Comment (ramiro050) [inline lib/Conversion/TorchToMhlo/Basic.cpp:1497]: Can you add an e2e test that passes on the TOSA backend?

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -3994,6 +3996,8 @@ class ConvertTorchToTosa : public ConvertTorchToTosaBase<ConvertTorchToTosa> {
     INSERT_BINARY_COMPARE_PATTERN(AtenNeTensorOp, tosa::EqualOp)
     INSERT_BINARY_COMPARE_PATTERN(AtenNeScalarOp, tosa::EqualOp)
     INSERT_BINARY_COMPARE_PATTERN(AtenBitwiseAndTensorOp, tosa::BitwiseAndOp)
+    INSERT_BINARY_COMPARE_PATTERN(AtenBitwiseOrTensorOp, tosa::BitwiseOrOp)
+    INSERT_BINARY_COMPARE_PATTERN(AtenBitwiseXorTensorOp, tosa::BitwiseXorOp)
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4000]: Can you add an e2e test that passes in the TOSA backend?

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: test/Conversion/TorchToMhlo/elementwise.mlir [hunk 3]
```diff
@@ -626,6 +626,61 @@ func.func @torch.aten.div.Tensor_mode$floor(%arg0: !torch.vtensor<[?,?,?,?],f32>
...
 // -----
 
+// CHECK-LABEL:   func.func @torch.aten.bitwise_or$basic(
+// CHECK-SAME:      %[[ARG0:.*]]: !torch.vtensor<[?,?],si32>, %[[ARG1:.*]]: !torch.vtensor<[?,?],si32>) -> !torch.vtensor<[?,?],si32> {
+// CHECK:           %[[T0:.*]] = torch_c.to_builtin_tensor %[[ARG0]] : !torch.vtensor<[?,?],si32> -> tensor<?x?xi32>
+// CHECK:           %[[T1:.*]] = torch_c.to_builtin_tensor %[[ARG1]] : !torch.vtensor<[?,?],si32> -> tensor<?x?xi32>
+// CHECK:           %[[T2:.*]] = chlo.broadcast_or %[[T0]], %[[T1]] : (tensor<?x?xi32>, tensor<?x?xi32>) -> tensor<?x?xi32>
+// CHECK:           %[[T3:.*]] = torch_c.from_builtin_tensor %[[T2]] : tensor<?x?xi32> -> !torch.vtensor<[?,?],si32>
+// CHECK:           return %[[T3]] : !torch.vtensor<[?,?],si32>
+func.func @torch.aten.bitwise_or$basic(%arg0: !torch.vtensor<[?,?],si32>, %arg1: !torch.vtensor<[?,?],si32>) -> !torch.vtensor<[?,?],si32> {
```
Comment (ramiro050) [inline test/Conversion/TorchToMhlo/elementwise.mlir:636]: These lit tests and the ones for the TOSA backend are no longer needed. See: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

> DON'T use a unit test if your lowering pattern could be described as a trivial "macro expansion" of one op into another op or set of ops. That is, if you feel like your unit test is just rewriting b.create<...>(...) into CHECK: ... then it is probably not a useful unit test.

File: lib/Conversion/TorchToMhlo/Basic.cpp (hunk 1 above)
Comment (ramiro050) [inline lib/Conversion/TorchToMhlo/Basic.cpp:1497]: Is there a reason the new e2e test added does not pass in MHLO?

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToMhlo/Basic.cpp (hunk 1 above)
Comment (li-plus) [reply lib/Conversion/TorchToMhlo/Basic.cpp:1497]: It seems that in the case of dynamic shape, `chlo.broadcast_and` will generate a `shape.cstr_broadcastable` op that cannot be lowered to mhlo. The existing `bitwise_and` e2e test does not pass in MHLO & TOSA either.

```
debug.mlir:2:8: error: failed to legalize operation 'shape.cstr_broadcastable'
  %0 = torch.aten.bitwise_and.Tensor %arg0, %arg1 : !torch.vtensor<[?,?],si32>, !torch.vtensor<[?,?],si32> -> !torch.vtensor<[?,?],si32>
       ^
debug.mlir:2:8: note: see current operation: %2 = "shape.cstr_broadcastable"(%0, %1) : (tensor<2xindex>, tensor<2xindex>) -> !shape.witness
debug.mlir:0:0: error: Module does not conform to the MHLO backend contract. See dialect conversion legality information above.
// -----// IR Dump After VerifyMhloBackendContract Failed (torch-verify-mhlo-backend-contract) //----- //
module {
  func.func @torch.aten.bitwise_and$basic(%arg0: tensor<?x?xi32>, %arg1: tensor<?x?xi32>) -> tensor<?x?xi32> {
    %0 = shape.shape_of %arg0 : tensor<?x?xi32> -> tensor<2xindex>
    %1 = shape.shape_of %arg1 : tensor<?x?xi32> -> tensor<2xindex>
    %2 = shape.cstr_broadcastable %0, %1 : tensor<2xindex>, tensor<2xindex>
    %3 = shape.assuming %2 -> (tensor<?x?xi32>) {
      %4 = shape.broadcast %0, %1 : tensor<2xindex>, tensor<2xindex> -> tensor<2xindex>
      %5 = "mhlo.dynamic_broadcast_in_dim"(%arg0, %4) {broadcast_dimensions = dense<[0, 1]> : tensor<2xi64>} : (tensor<?x?xi32>, tensor<2xindex>) -> tensor<?x?xi32>
      %6 = "mhlo.dynamic_broadcast_in_dim"(%arg1, %4) {broadcast_dimensions = dense<[0, 1]> : tensor<2xi64>} : (tensor<?x?xi32>, tensor<2xindex>) -> tensor<?x?xi32>
      %7 = mhlo.and %5, %6 : tensor<?x?xi32>
      shape.assuming_yield %7 : tensor<?x?xi32>
    }
    return %3 : tensor<?x?xi32>
  }
}
```

File: test/Conversion/TorchToMhlo/elementwise.mlir (hunk 3 above)
Comment (li-plus) [reply test/Conversion/TorchToMhlo/elementwise.mlir]: Okay, got these tests removed.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (li-plus) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4005]: Okay, I added two e2e tests for each bitwise binary op, one with dynamic shape and the other with static shape. After fixing type promotion issue, now TOSA can pass both dynamic & static shape tests, and MHLO passes only the static one.

File: lib/Conversion/TorchToMhlo/Basic.cpp (hunk 1 above)
Comment (li-plus) [reply lib/Conversion/TorchToMhlo/Basic.cpp:1494]: Update: the existing `bitwise_and` e2e test passed in TOSA after fixing type promotion, but still failed in MHLO as expected.

**Review by ramiro050 — APPROVED** [review]
Thanks for the fixes! In general, every time op is added to any backend, a test using the new changes should pass e2e. The same applies to this PR https://github.com/llvm/torch-mlir/pull/1752 that got merged. Please make a new PR adding tests that pass on the MHLO side

**Comment by li-plus** [comment]
@ramiro050 I have rebased to the latest HEAD and resolved conflicts. Would you please check if those conversations could be resolved and merge this PR?

**Review by ramiro050 — APPROVED** [review]

**Pull request merged**